src/snowflake/cli/api/connections.py

```diff
@@ -0,0 +1,349 @@
+        return self._diag_allowlist_path
+
+    def set_diag_allowlist_path(self, value: Optional[Path]):
+        self._diag_allowlist_path = value
+
+    @property
+    def temporary_connection(self) -> bool:
+        return self._temporary_connection
+
+    def set_temporary_connection(self, value: bool):
+        self._temporary_connection = value
+
+    @property
+    def session_token(self) -> Optional[str]:
+        return self._session_token
+
+    def set_session_token(self, value: Optional[str]):
+        self._session_token = value
+
+    @property
+    def master_token(self) -> Optional[str]:
+        return self._master_token
+
+    def set_master_token(self, value: Optional[str]):
+        self._master_token = value
+
+    @property
+    def token_file_path(self) -> Optional[Path]:
+        return self._token_file_path
+
+    def set_token_file_path(self, value: Optional[Path]):
+        self._token_file_path = value
+
+    def _collect_not_empty_connection_attributes(self):
+        return {
+            "account": self.account,
+            "user": self.user,
+            "password": self.password,
+            "authenticator": self.authenticator,
+            "private_key_path": self.private_key_path,
+            "database": self.database,
+            "schema": self.schema,
+            "role": self.role,
+            "warehouse": self.warehouse,
+            "session_token": self.session_token,
+            "master_token": self.master_token,
+            "token_file_path": self.token_file_path,
+        }
+
+    def build_connection(self):
+        from snowflake.cli._app.snow_connector import connect_to_snowflake
+
+        # Ignore warnings about bad owner or permissions on Windows
+        # Telemetry omit our warning filter from config.py
+        if IS_WINDOWS:
+            warnings.filterwarnings(
+                action="ignore",
+                message="Bad owner or permissions.*",
+                module="snowflake.connector.config_manager",
+            )
+
+        return connect_to_snowflake(
+            temporary_connection=self.temporary_connection,
+            mfa_passcode=self._mfa_passcode,
+            enable_diag=self._enable_diag,
+            diag_log_path=self._diag_log_path,
+            diag_allowlist_path=self._diag_allowlist_path,
+            connection_name=self.connection_name,
+            **self._collect_not_empty_connection_attributes(),
+        )
+
+
+class OpenConnectionCache:
+    """
+    A connection cache that transparently manages SnowflakeConnection objects
+    and is keyed by ConnectionContext objects, e.g. cache[ctx].execute_string(...).
+    Connections are automatically closed after CONNECTION_CLEANUP_SEC, but
+    are guaranteed to be open (if config is valid) when returned by the cache.
+    """
+
+    connections: dict[str, SnowflakeConnection]
+    cleanup_futures: dict[str, asyncio.TimerHandle]
+
+    CONNECTION_CLEANUP_SEC: float = 10.0 * 60
+
```

Comment:
No, we don't. It's something I'm looking for CLI's feedback on (@sfc-gh-turbaszek) -- also @sfc-gh-klin would be helpful to get your thoughts too. This timeout says that if it's been 10 minutes since a connection was accessed, we will attempt to close it.

This auto-close will _not_ cancel any in-fight queries, as far as I can tell.

Reply:
In the VS Code extension by default we keep our connection through the Node JS connector open the whole time the user is using the extension unless they change a config setting. But that continuous extension usage connection (to allow users to come back to their session and execute SQL or browse their data) is different to how the connections here are going to be used. Since these connections are being used for execution of SnowCLI commands, if it has been 10 minutes since one has been used, it seems good to close it to save resources. When the user from the extension tries to execute another SnowCLI command, opening a new one or re-using an existing one doesn't make a difference from their perspective--maybe opening a new one takes a bit longer than re-using an existing one but that overhead doesn't seem bad. So this is reasonable.

Reply:
It's possible to drop the connection cache altogether if you don't think the performance gains are worth the complexity. It's not a correctness thing in this PR, because the session / master token are passed in anyway to ensure the "identity" of the connection is shared between LSP/CLI + Vscode.